crates/starknet-types-core/src/felt/qm31.rs

```diff
@@ -0,0 +1,487 @@
+use core::fmt;
+
+use crate::felt::Felt;
+
+pub const STWO_PRIME: u64 = (1 << 31) - 1;
+const STWO_PRIME_U128: u128 = STWO_PRIME as u128;
+const MASK_36: u64 = (1 << 36) - 1;
+const MASK_8: u64 = (1 << 8) - 1;
+
+#[derive(Debug)]
+pub enum QM31Error {
+    UnreducedFelt(Felt),
+    FeltTooBig(Felt),
+    InvalidInversion,
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for QM31Error {}
+
+impl fmt::Display for QM31Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            QM31Error::UnreducedFelt(felt) => writeln!(
+                f,
+                "Number is not a packing of a QM31 in reduced form: {felt})"
+            ),
+            QM31Error::FeltTooBig(felt) => writeln!(
+                f,
+                "Number used as QM31 since it's more than 144 bits long: {felt}"
+            ),
+            QM31Error::InvalidInversion => writeln!(f, "Attempt to invert a qm31 equal to zero"),
```

Comment:
I don't think `writeln` should be used inside `fmt` impl.
It's the caller responsibility to decide if he want to add a newline.